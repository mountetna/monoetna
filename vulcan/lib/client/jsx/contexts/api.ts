import {useCallback, useRef, useState} from 'react';
import {showMessages} from 'etna-js/actions/message_actions';
import {
  checkStatus,
  handleFetchError,
  handleFetchSuccess,
  headers
} from 'etna-js/utils/fetch';
import {
  defaultSessionStatusResponse,
  defaultWorkflowsResponse,
  FiguresResponse,
  SessionStatusResponse,
  VulcanFigureSession,
  VulcanSession,
  WorkflowsResponse
} from '../api_types';

export const defaultApiHelpers = {
  showErrors<T>(work: Promise<T>): Promise<T> {
    work.catch((e) => {
      console.error(e);
    });

    return work;
  },
  getData(url: string): Promise<any> {
    return new Promise(() => null);
  },
  postInputs(session: VulcanSession): Promise<SessionStatusResponse> {
    return new Promise(() => null);
  },
  pollStatus(session: VulcanSession): Promise<SessionStatusResponse> {
    return new Promise(() => null);
  },
  getWorkflows(): Promise<WorkflowsResponse> {
    return new Promise(() => null);
  },
  fetchFigures(projectName: string): Promise<FiguresResponse> {
    return new Promise(() => null);
  },
  fetchFigure(
    projectName: string,
    figureId: number
  ): Promise<VulcanFigureSession> {
    return new Promise(() => null);
  },
  createFigure(projectName: string, params: any): Promise<VulcanFigureSession> {
    return new Promise(() => null);
  },
  updateFigure(projectName: string, params: any): Promise<VulcanFigureSession> {
    return new Promise(() => null);
  },
  deleteFigure(
    projectName: string,
    figureId: number
  ): Promise<VulcanFigureSession> {
    return new Promise(() => null);
  },
  updateFigureDependencies(projectName: string, figureId: number): Promise<VulcanFigureSession> {
    return new Promise(() => null);
  }
};

export function useApi(
  invoke: (a: {type: string}) => any
): typeof defaultApiHelpers {
  const vulcanPath = useCallback(
    (endpoint: string) => `${CONFIG.vulcan_host}${endpoint}`,
    []
  );
  const vulcanPost = useCallback((endpoint: string, params: Object) => {
    return fetch(endpoint, {
      method: 'POST',
      credentials: 'include',
      headers: headers('json'),
      body: JSON.stringify({
        ...params
      })
    }).then(checkStatus);
  }, []);

  const rawVulcanGet = useCallback((endpoint: string) => {
    return fetch(endpoint, {
      method: 'GET',
      credentials: 'include',
      headers: headers('json')
    });
  }, []);

  const vulcanDelete = useCallback((endpoint: string) => {
    return fetch(endpoint, {
      method: 'DELETE',
      credentials: 'include',
      headers: headers('json')
    });
  }, []);

  const vulcanGet = useCallback(
    (endpoint: string) => {
      return rawVulcanGet(endpoint)
        .then(checkStatus)
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [rawVulcanGet]
  );

  const getWorkflows = useCallback((): Promise<WorkflowsResponse> => {
    // old: ROUTES.fetch_workflows
    return vulcanGet(vulcanPath('/api/workflows'))
      .then(handleFetchSuccess)
      .catch(handleFetchError);
  }, [vulcanGet, vulcanPath]);

  const postInputs = useCallback(
    (session: VulcanSession): Promise<SessionStatusResponse> => {
      if (!session.workflow_name) {
        return Promise.reject(
          new Error('No workflow selected, bug in client.')
        );
      }

      return vulcanPost(
        // old: ROUTES.submit
        vulcanPath(`/api/${session.project_name}/session/${workflow_name}`),
        session
      )
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [vulcanPath, vulcanPost]
  );

  const pollStatus = useCallback(
    (session: VulcanSession): Promise<SessionStatusResponse> => {
      if (!session.workflow_name) {
        return Promise.reject(
          new Error('No workflow selected, bug in client.')
        );
      }

      return vulcanPost(
        // Old: ROUTES.status
        vulcanPath(`/api/${session.project_name}/session/${session.workflow_name}/status`),
        session
      )
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [vulcanPath, vulcanPost]
  );

  const getData = useCallback(
    (url: string) => {
      return vulcanGet(url)
        .then(handleFetchSuccess)
        .catch(handleFetchError)
        .then((data) => {
          // TODO: In the future, we should set content type headers to inform the client, for now we aggressively
          // try to parse JSON
          try {
            return JSON.parse(data);
          } catch {
            return data;
          }
        });
    },
    [vulcanGet]
  );

  const showErrors = useCallback(
    <T>(work: Promise<T>): Promise<T> => {
      work.catch((e) => {
        if (!(e instanceof Array)) {
          e = [`${e}`];
        }

        console.error(e);
        invoke(showMessages(e));
      });

      return work;
    },
    [invoke]
  );

  const fetchFigures = useCallback(
    (projectName: string): Promise<FiguresResponse> => {
      // Old: ROUTES.fetch_figures
      return vulcanGet(vulcanPath(`/api/${projectName}/figures`))
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [vulcanGet, vulcanPath]
  );

  const fetchFigure = useCallback(
    (projectName: string, figureId: number): Promise<VulcanFigureSession> => {
      // Old: ROUTES.fetch_figure
      return vulcanGet(vulcanPath(`/api/${projectName}/figure/${figureId}`))
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [vulcanGet, vulcanPath]
  );

  const updateFigure = useCallback(
    (projectName: string, params: any): Promise<VulcanFigureSession> => {
      return vulcanPost(
        // Old: ROUTES.update_figure
        vulcanPath(`/api/${projectName}/figure/${params.figure_id}/update`),
        params
      )
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [vulcanPost, vulcanPath]
  );

  const updateFigureDependencies = useCallback(
    (projectName: string, figure_id: number): Promise<VulcanFigureSession> => {
      return vulcanPost(
        // Old: ROUTES.update_figure
        vulcanPath(`/api/${projectName}/figure/${figure_id}/update`),
        {
          update_dependencies: true,
          comment: 'Updating dependencies and snapshot'
        }
      )
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [vulcanPost, vulcanPath]
  );

  const createFigure = useCallback(
    (projectName: string, params: any): Promise<VulcanFigureSession> => {
      // Old: ROUTES.create_figure
      return vulcanPost(vulcanPath(`/api/${projectName}/figure/create`), params)
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [vulcanPost, vulcanPath]
  );

  const deleteFigure = useCallback(
    (projectName: string, figureId: number): Promise<VulcanFigureSession> => {
      return vulcanDelete(
        // Old: ROUTES.delete_figure
        vulcanPath(`/api/${projectName}/figure/${figureId}`)
      )
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [vulcanPath, vulcanDelete]
  );

  return {
    showErrors,
    getData,
    getWorkflows,
    postInputs,
    pollStatus,
    deleteFigure,
    createFigure,
    updateFigure,
    fetchFigures,
    fetchFigure,
    updateFigureDependencies
  };
}
