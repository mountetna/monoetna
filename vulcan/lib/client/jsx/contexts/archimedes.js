import React, {useReducer, createContext} from 'react';
import {
  SET_DATA,
  SET_WORKFLOW,
  SET_WORKFLOWS,
  SET_STATUS
} from '../actions/archimedes';
import ArchimedesReducer from '../reducers/archimedes';

export const ArchimedesContext = createContext();

export const ArchimedesProvider = (props) => {
  const initialState = {
    workflows: {},
    workflow: {}
  };

  const [state, dispatch] = useReducer(ArchimedesReducer, initialState);

  const setWorkflows = (workflows) => {
    dispatch({type: SET_WORKFLOWS, workflows});
  };

  const setWorkflow = (workflow) => {
    dispatch({type: SET_WORKFLOW, workflow});
  };

  const setStatus = (status) => {
    dispatch({type: SET_STATUS, status});
  };

  const setData = (url, data) => {
    dispatch({type: SET_DATA, url, data});
  };

  return (
    <ArchimedesContext.Provider
      value={{
        workflows: state.workflows,
        workflow: state.workflow,
        setWorkflows,
        setWorkflow,
        setData,
        setStatus
      }}
    >
      {props.children}
    </ArchimedesContext.Provider>
  );
};
