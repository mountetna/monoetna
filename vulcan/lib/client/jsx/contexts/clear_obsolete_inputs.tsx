import {VulcanState} from "../reducers/vulcan_reducer";
import {Dispatch, useEffect, useRef} from "react";
import * as _ from 'lodash';
import {setInputs, VulcanAction} from "../actions/vulcan";
import {SessionStatusResponse} from "../api_types";

export function useClearObsoleteInputs(state: VulcanState, dispatch: Dispatch<VulcanAction>) {
    const {inputs, status, workflow} = state;
    const lastInputsRef = useRef({inputs});

    useEffect(() => {
        if (workflow == null) return;
        const {dependencies_of_outputs} = workflow;

        const {current: {inputs: lastInputs}} = lastInputsRef;

        let newInputs: SessionStatusResponse['session']['inputs'] = inputs;

        function clearDependentInputs(source: string) {
            const dependencies = dependencies_of_outputs[source];
            if (dependencies) {
                dependencies.forEach(source => {
                    if (source in newInputs) {
                        if (newInputs === inputs) newInputs = {...inputs};
                        delete newInputs[source];
                    }
                });
            }
        }

        Object.keys(inputs).forEach(source => {
            if (source in lastInputs && !_.isEqual(inputs[source], lastInputs[source])) {
                clearDependentInputs(source);
            }
        });

        if (newInputs !== inputs) {
            dispatch(setInputs(newInputs));
        }

        lastInputsRef.current.inputs = inputs;
    }, [inputs, status, workflow]);
}