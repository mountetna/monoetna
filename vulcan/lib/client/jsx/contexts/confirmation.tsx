import {Cancellable} from "etna-js/utils/cancellable";
import {useCallback} from "react";

export const defaultConfirmationHelpers = {
  confirm(message: string, context: Cancellable) {
    return Promise.resolve(true)
  },
}

export function useConfirmation(windowConfirm = window.confirm): typeof defaultConfirmationHelpers {
  const openConfirmationModal = useCallback((message: string): [Promise<boolean>, Function] => {
    // TODO: Implement some proper modal that allows for closing via the returned callback.
    return [Promise.resolve(windowConfirm(message)), () => {
      // This callback should close the modal if displaying a truly async modal.
      // Used to cleanup when the modal needs to be forced closed.
    }]
  }, [windowConfirm]);

  const confirm = useCallback((message: string, context: Cancellable) => {
    const [promise, cleanup] = openConfirmationModal(message);
    return context.race(promise).then(({ cancelled, result }) => {
      if (cancelled || typeof result === "undefined") {
        cleanup();
        return false;
      }

      return result;
    })
  }, [openConfirmationModal])

  return {
    confirm,
  }
}