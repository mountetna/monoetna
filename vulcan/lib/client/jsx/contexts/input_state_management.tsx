import {VulcanState} from "../reducers/vulcan_reducer";
import {Dispatch, useEffect, useMemo} from "react";
import {patchInputs, removeDownloads, removeInputs, VulcanAction} from "../actions/vulcan";
import {
  isPendingUiQuery,
  missingUiQueryOutputs,
  pendingSteps, sourceNameOfReference, statusOfStep, stepOfSource
} from "../selectors/workflow_selectors";

export function useInputStateManagement(state: VulcanState, dispatch: Dispatch<VulcanAction>, statusIsFresh: boolean) {
  const {inputs, workflow, status, data, session} = state;

  useEffect(() => {
    if (workflow == null) return;
    if (!statusIsFresh) return;
    const inputDeletes: {[k: string]: true} = {};
    const downloadDeletes: {[k: string]: true} = {};
    const stepsWithDownloads: {[k: string]: true} = {};

    const droppedSteps = workflow.steps[0].filter(step => {
      const stepStatus = statusOfStep(step, status);
      if (step.out.length === 0) return false;

      if (!stepStatus) return true;

      return !step.out.every(outName => {
        const source = sourceNameOfReference([step.name, outName]);
        if (source in session.inputs) {
          return true;
        }

        if (!stepStatus.downloads) return false;
        stepsWithDownloads[step.name] = true;

        return outName in stepStatus.downloads;
      });
    });

    droppedSteps.forEach(step => {
      step.out.forEach(outName => {
        const source = sourceNameOfReference([step.name, outName]);
        if (source in session.inputs) inputDeletes[source] = true;
        if (step.name in stepsWithDownloads) downloadDeletes[step.name] = true;

        workflow.dependencies_of_outputs[source].forEach(dependent => {
          const stepName = stepOfSource(dependent);
          if (dependent in inputs) inputDeletes[dependent] = true;
          if (stepName && stepName in stepsWithDownloads) downloadDeletes[stepName] = true;
        });
      })
    })

    let d = Object.keys(inputDeletes);
    if (d.length > 0) dispatch(removeInputs(d));

    d = Object.keys(downloadDeletes);
    if (d.length > 0) dispatch(removeDownloads(d));
  }, [inputs, workflow, status, data, session.inputs, statusIsFresh]);

  // We inject a `null` input into the session,
  //   to indicate that we're waiting for a user input value
  //   to return to the server.
  useEffect(() => {
    if (!workflow) return;

    let newInputs = {};
    let nextUiSteps = pendingSteps(workflow, status).filter(({step}) => isPendingUiQuery(step, status, data, session));

    nextUiSteps.forEach((nextInputStep) => {
      let missingInputs = missingUiQueryOutputs(nextInputStep.step, inputs);

      if (Object.keys(missingInputs).length > 0) {
        // Make sure to copy over the current inputs, otherwise
        //   they'll get wiped out in the reducer.
        newInputs = {
          ...newInputs,
          ...missingInputs,
        };
      }
    });

    if (Object.keys(newInputs).length > 0) dispatch(patchInputs(newInputs));
  }, [inputs, workflow, status, data, session]);
}