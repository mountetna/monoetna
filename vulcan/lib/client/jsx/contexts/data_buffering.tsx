import {Dispatch} from 'react';
import {setDownloadedData, VulcanAction} from '../actions/vulcan_actions';
import {defaultApiHelpers} from './api';
import {VulcanState} from '../reducers/vulcan_reducer';
import {shouldDownloadStep} from '../selectors/workflow_selectors';
import {useAsync} from 'etna-js/utils/cancellable_helpers';
import {runAttempts} from 'etna-js/utils/retryable';
import { MultiFileContentResponse, OutputConfig } from '../api_types';

export function useDataBuffering(
    state: VulcanState,
    dispatch: Dispatch<VulcanAction>,
    showErrors: typeof defaultApiHelpers.showErrors,
    readFiles: typeof defaultApiHelpers.readFiles,
) {
  const {status, workflow, data, projectName, workspaceId} = state;

  useAsync(function* () {
    if (!workspaceId) return;

    // Find the next status with a download that is an input to a ui-output or ui-query,
    // initiate a download, and let it run.
    for (let [step, stepStatus] of Object.entries(status.steps)) {
      if (stepStatus.status=="complete" && stepStatus.outputs?.files) {
        for (let outputName of stepStatus.outputs.files) {
          if (outputName in Object.keys(data)) continue;
          if (!shouldDownloadStep(stepStatus.name, workflow, outputName)) continue;

          const downloaded: MultiFileContentResponse = yield* runAttempts(() => readFiles(projectName, workspaceId, [outputName]));

          dispatch(setDownloadedData(outputName, downloaded[0]['content']));
          return;
        }
      }

    }
  }, [status, workflow, data]);
}
