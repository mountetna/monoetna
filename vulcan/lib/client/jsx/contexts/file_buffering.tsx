import {Dispatch} from 'react';
import {updateFiles, VulcanAction} from '../actions/vulcan_actions';
import {defaultApiHelpers} from './api';
import {VulcanState} from '../reducers/vulcan_reducer';
import {allFilesToBuffer, filesReturnToMultiFileContent} from '../selectors/workflow_selectors';
import {useAsync} from 'etna-js/utils/cancellable_helpers';
import { MultiFileContent } from '../api_types';

const imageExtRegEx = /\.(png|tiff?|jpe?g|bmp|svg|gif|webp)$/

export function useFileBuffering(
    state: VulcanState,
    dispatch: Dispatch<VulcanAction>,
    showErrors: typeof defaultApiHelpers.showErrors,
    getFileNames: typeof defaultApiHelpers.getFileNames,
    readFiles: typeof defaultApiHelpers.readFiles,
) {
  const {update_files, workspace, workspaceId, projectName} = state;

  useAsync(function* () {
    if (update_files && !!workspaceId && !!projectName) {
      console.log("Using data buffering")
      let update = {output_files: [] as string[], file_contents: {} as {[k: string]: any}};
      showErrors(getFileNames(projectName, workspaceId))
      .then((fileNamesRaw) => {
        const fileNames = fileNamesRaw.files;
        update['output_files'] = fileNames;
        let filesContent: MultiFileContent = {};
          
        const filesReady = allFilesToBuffer(workspace).filter(f => fileNames.includes(f));
        if (filesReady.length > 0) {
          // Stub images
          const images = filesReady.filter(f => imageExtRegEx.test(f));
          if (images.length > 0) {
            for (let ind in images) {
              filesContent[images[ind]] = '__IMAGE_CONTENT_STUB__'
            }
          }
          // Grab non-stubbed files
          const files = filesReady.filter(f => !Object.keys(filesContent).includes(f));
          if (files.length > 0) {
            showErrors(readFiles(projectName, workspaceId, files))
            .then((filesContentRaw) => {
              update['file_contents'] = {
                ...update['file_contents'],
                ...filesReturnToMultiFileContent(filesContentRaw)
              }
              dispatch(updateFiles(update))
            })
          } else {
            dispatch(updateFiles(update))
          }
        }
      })
    }
  }, [update_files, workspaceId, projectName]);
}
