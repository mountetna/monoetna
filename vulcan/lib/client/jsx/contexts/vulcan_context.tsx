import React, {useReducer, createContext, useRef, useState, useCallback} from 'react';
import VulcanReducer, {defaultVulcanState, VulcanState} from '../reducers/vulcan_reducer';
import {VulcanAction} from '../actions/vulcan_actions';
import {defaultSessionStorageHelpers, useLocalSessionStorage} from './session_storage';
import {defaultApiHelpers, useApi} from './api';
import {defaultSessionSyncHelpers, useSessionSyncWhileRunning} from './session_sync_while_running';
import {defaultInputStateManagement, useInputStateManagement} from './input_state_management';
import {useActionInvoker} from 'etna-js/hooks/useActionInvoker';
import {defaultConfirmationHelpers, useConfirmation} from './confirmation';
import { useWorkspaceWorkflowLoading } from './workspace_worksflow_loading';
import { useFileBuffering } from './file_buffering';

export const defaultContext = {
  state: defaultVulcanState as VulcanState,
  stateRef: {current: defaultVulcanState}, // This would be set with a context dispatch when the provide is actually
                                           // installed.
  dispatch: (a: VulcanAction) => console.warn('action dispatched but not handled', a),
  useActionInvoker: (() => () => null) as typeof useActionInvoker, ...defaultConfirmationHelpers, ...defaultSessionStorageHelpers, ...defaultApiHelpers, ...defaultSessionSyncHelpers, ...defaultInputStateManagement,
};

export type VulcanContextData = typeof defaultContext;
export const VulcanContext = createContext(defaultContext);
export type VulcanContext = typeof VulcanContext;
export type ProviderProps = {
  params?: {},
  storage?: typeof localStorage,
  logActions?: boolean,
  wrapper?: [Function, Function],
  children: any
};

// Existing mostly to down type the return result into A from potentially inferring A & B
function withOverrides<A, B extends Partial<A>>(base: A, overrides: B): A {
  const result: A = {...base};
  for (let k in overrides) {
    const v = overrides[k];
    if (v !== undefined) { // @ts-ignore
      result[k] = v;
    }
  }

  return result;
}

export const VulcanProvider = (props: ProviderProps & Partial<VulcanContextData>) => {
  const [start, end] = props.wrapper || [() => null, () => null];
  start();

  try {
    const actionInvokerHelpers = withOverrides({useActionInvoker}, props);
    const invoker = actionInvokerHelpers.useActionInvoker();
    const stateRef = useRef(props.state || defaultContext.state);
    const [state, setState] = useState(stateRef.current);

    const dispatch = useCallback((action: VulcanAction) => {
      if (props.logActions) {
        console.log(action.type, action);
      }

      const result = VulcanReducer(stateRef.current, action);
      stateRef.current = result;
      setState(result);

      return result;
    }, [props.logActions]);

    if (!!props.params && 'project_name' in props.params && state.projectName!=props.params.project_name) {
      const newResult = {...stateRef.current, projectName: props.params.project_name as string};
      stateRef.current = newResult;
      setState(newResult);
    }
    if (!!props.params && 'workspace_id' in props.params && state.workspaceId!=props.params.workspace_id) {
      const newResult = {...stateRef.current, workspaceId: Number(props.params.workspace_id) as number};
      stateRef.current = newResult;
      setState(newResult);
    }

    const localSessionHelpers = withOverrides(useLocalSessionStorage(state, props), props);
    const apiHelpers = withOverrides(useApi(invoker), props);
    const {showErrors, readFiles, getWorkflows, getWorkspaces, getWorkspace, requestRun, pullRunStatus, postUIValues, getFileNames} = apiHelpers;
    const sessionSyncHelpers = withOverrides(
      useSessionSyncWhileRunning(showErrors, requestRun, pullRunStatus, postUIValues, dispatch),
      props
    );
    useWorkspaceWorkflowLoading(JSON.stringify(props.params), dispatch, getWorkflows, getWorkspaces, getWorkspace, showErrors, state.projectName, state.workspaceId, state.update_workflows);
    useFileBuffering(state, dispatch, showErrors, getFileNames, readFiles)
    const confirmationHelpers = withOverrides(useConfirmation(), props);
    const inputHelpers = useInputStateManagement(invoker, dispatch, sessionSyncHelpers.requestPoll, stateRef);

    return (<VulcanContext.Provider value={{
      state,
      stateRef,
      dispatch,
      ...confirmationHelpers,
      ...actionInvokerHelpers,
      ...localSessionHelpers,
      ...apiHelpers,
      ...sessionSyncHelpers,
      ...inputHelpers,
    }}>
      {props.children}
    </VulcanContext.Provider>);
  } finally {
    end();
  }
};
