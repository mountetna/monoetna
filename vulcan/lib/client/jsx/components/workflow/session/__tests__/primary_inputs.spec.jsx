import React from 'react';
import {VulcanProvider} from '../../../../contexts/vulcan_context';
import renderer from 'react-test-renderer';
import PrimaryInputs from '../primary_inputs';

describe('PrimaryInputs', () => {
  let state;

  beforeEach(() => {
    state = {
      workflow: {
        inputs: {
          anInt: {
            type: 'int',
            default: 1,
            doc: 'help',
            group: 'one'
          },
          aFloat: {
            type: 'float',
            default: 1.2,
            doc: 'help',
            group: 'one'
          },
          aBool: {
            type: 'boolean',
            default: true,
            doc: 'help',
            group: 'one'
          },
          anIntWithoutDefault: {
            type: 'int',
            default: null,
            doc: 'help',
            group: 'two'
          },
          aFloatWithoutDefault: {
            type: 'float',
            default: null,
            doc: 'help',
            group: 'two'
          },
          aBoolWithoutDefault: {
            type: 'boolean',
            default: null,
            doc: 'help',
            group: 'two'
          }
        }
      },
      session: {
        key: 'session_key'
      }
    };
  });

  it('renders each type of input', () => {
    // Wrap with Provider here so store gets passed down to child components in Context
    const component = renderer.create(
      <VulcanProvider state={state}>
        <PrimaryInputs />
      </VulcanProvider>
    );

    let instance = component.root;

    expect(instance.findAllByType('input').length).toEqual(
      Object.keys(state.workflow.inputs).length
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
