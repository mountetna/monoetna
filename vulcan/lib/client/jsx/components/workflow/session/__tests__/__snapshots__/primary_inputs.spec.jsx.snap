// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrimaryInputs renders each type of input 1`] = `
<div
  className="primary-inputs inputs-pane"
>
  <div
    className="title"
  >
    Inputs
  </div>
  <div
    className="primary-inputs-container items"
  >
    <div
      className="view_item"
    >
      <div
        className="item_name"
      >
        anInt
      </div>
      <div
        className="item_view"
      >
        <input
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value={1}
        />
      </div>
    </div>
    <div
      className="view_item"
    >
      <div
        className="item_name"
      >
        aFloat
      </div>
      <div
        className="item_view"
      >
        <input
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value={1.2}
        />
      </div>
    </div>
    <div
      className="view_item"
    >
      <div
        className="item_name"
      >
        aBool
      </div>
      <div
        className="item_view"
      >
        <input
          className="text_box"
          defaultChecked={true}
          onChange={[Function]}
          type="checkbox"
        />
      </div>
    </div>
    <div
      className="view_item"
    >
      <div
        className="item_name"
      >
        anIntWithoutDefault
      </div>
      <div
        className="item_view"
      >
        <input
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="view_item"
    >
      <div
        className="item_name"
      >
        aFloatWithoutDefault
      </div>
      <div
        className="item_view"
      >
        <input
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="view_item"
    >
      <div
        className="item_name"
      >
        aBoolWithoutDefault
      </div>
      <div
        className="item_view"
      >
        <input
          className="text_box"
          defaultChecked={null}
          onChange={[Function]}
          type="checkbox"
        />
      </div>
    </div>
  </div>
</div>
`;
